src/gc.c

```diff
@@ -2834,9 +2864,23 @@ void gc_mark_loop_parallel(jl_ptls_t ptls, int master)
         gc_mark_and_steal(ptls);
         jl_atomic_fetch_add(&gc_n_threads_marking, -1);
     }
-    while (jl_atomic_load(&gc_n_threads_marking) > 0) {
+    while (1) {
+        uv_mutex_lock(&gc_mark_loop_gate_lock);
+        count_work : {
+            int n_threads_marking = jl_atomic_load(&gc_n_threads_marking);
+            if (n_threads_marking == 0) {
+                uv_mutex_unlock(&gc_mark_loop_gate_lock);
+                break;
+            }
+            int64_t work = gc_estimate_mark_work();
+            if (work <= n_threads_marking * GC_MIN_MARK_WORK_PER_THREAD) {
+                jl_cpu_pause();
+                goto count_work;
```

Comment:
So, the first thread to get into this loop holds the lock and the remaining threads sleep on trying to acquire the lock? Is that the intention?

Reply:
Having a comment that explains the logic would be great!

Reply:
That's the intention. We should have only one thread counting the available work and only one GC thread entering the mark-loop at a time.

Will add some comments to clarify that.

Reply:
It might be more efficient to avoid the need for every thread to acquire the lock to estimate the work. If you use the `trylock` variant (I think there is one...), you can have one thread win the race to the lock and estimate the work, while the threads that lose the race can block on a condition instead.